pinot-core/src/main/java/org/apache/pinot/core/util/TlsUtils.java

```diff
@@ -0,0 +1,282 @@
+   * store paths and passwords
+   * @param keyStoreType keystore type for keystore
+   * @param keyStorePath key store path
+   * @param keyStorePassword key password
+   * @param trustStoreType keystore type for truststore
+   * @param trustStorePath trust store path
+   * @param trustStorePassword trust password
+   */
+  public static void installDefaultSSLSocketFactory(String keyStoreType, String keyStorePath, String keyStorePassword,
+      String trustStoreType, String trustStorePath, String trustStorePassword) {
+    KeyManager[] keyManagers = null;
+    if (keyStorePath != null) {
+      keyManagers = createKeyManagerFactory(keyStorePath, keyStorePassword, keyStoreType).getKeyManagers();
+    }
+
+    TrustManager[] trustManagers = null;
+    if (trustStorePath != null) {
+      trustManagers = createTrustManagerFactory(trustStorePath, trustStorePassword, trustStoreType).getTrustManagers();
+    }
+
+    try {
+      SSLContext sc = SSLContext.getInstance("SSL");
+      sc.init(keyManagers, trustManagers, new java.security.SecureRandom());
+
+      // HttpsURLConnection
+      HttpsURLConnection.setDefaultSSLSocketFactory(sc.getSocketFactory());
+
+      // Apache HTTP client 3.x
+      Protocol.registerProtocol("https",
+          new Protocol(CommonConstants.HTTPS_PROTOCOL, new PinotProtocolSocketFactory(sc.getSocketFactory()), 443));
+
+      // FileUploadDownloadClient
+      FileUploadDownloadClient.installDefaultSSLContext(sc);
+    } catch (GeneralSecurityException e) {
+      throw new IllegalStateException("Could not initialize SSL support", e);
+    }
+  }
+
+  private static String key(String namespace, String suffix) {
+    return namespace + "." + suffix;
+  }
+
+  /**
+   * Adapted from: https://svn.apache.org/viewvc/httpcomponents/oac
+   * .hc3x/trunk/src/contrib/org/apache/commons/httpclient/contrib/ssl/AuthSSLProtocolSocketFactory.java?view=markup
+   */
+  private static class PinotProtocolSocketFactory implements ProtocolSocketFactory {
+    final SSLSocketFactory _sslSocketFactory;
+
+    public PinotProtocolSocketFactory(SSLSocketFactory sslSocketFactory) {
+      _sslSocketFactory = sslSocketFactory;
+    }
+
+    @Override
+    public Socket createSocket(String host, int port, InetAddress localAddress, int localPort)
+        throws IOException, UnknownHostException {
+      return _sslSocketFactory.createSocket(host, port, localAddress, localPort);
+    }
+
+    @Override
+    public Socket createSocket(String host, int port, InetAddress localAddress, int localPort,
+        HttpConnectionParams params)
+        throws IOException, UnknownHostException, ConnectTimeoutException {
+      Preconditions.checkNotNull(params);
+
+      int timeout = params.getConnectionTimeout();
+      if (timeout <= 0) {
+        return _sslSocketFactory.createSocket(host, port, localAddress, localPort);
+      }
+
+      Socket socket = _sslSocketFactory.createSocket();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *socket*. The resource is referenced at lines 271, 272. The resource is returned at line 273. There are other execution paths that don&#39;t close the resource or return it, for example, when Socket.connect throws an exception. To prevent this resource leak, close *socket* along these other paths before you exit this method.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).